In [1]:
####Need to bypass SSL verification therefore a custom http client is needed#################
from arango.http import HTTPClient
from arango.response import Response
from requests import Session

import warnings
warnings.simplefilter('ignore')

class CustomHTTPClient(HTTPClient):
    def __init__(self):
        self._session = Session()
    
    def send_request(self, method, url,  params=None,    data=None,    headers=None,    auth=None):
   
        response = self._session.request(
        method=method,
        url=url,
        params=params,
        data=data,
        headers=headers,
        auth=auth,
        verify=False # Disable SSL verification
        )
        
        return Response(
        method=response.request.method,
        url=response.url,
        headers=response.headers,
        status_code=response.status_code,
        status_text=response.reason,
        raw_body=response.text,
        )
#######################################################################################################

import pandas as pd
import boto3
import os
from arango import ArangoClient

# S3 credential  SIddhant's credential
#s3session = boto3.Session(aws_access_key_id="AKIAJUGBG25TI5UIRPLA", aws_secret_access_key="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")
#s3resource = s3session.resource('s3')
#s3client = s3session.client('s3')

In [17]:
# using Tina's credential  for S3
s3session = boto3.Session(aws_access_key_id="AKIAJ2K3S7ZZHJIS5KMQ", aws_secret_access_key="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")
s3resource = s3session.resource('s3')
s3client = s3session.client('s3')

# Initialize the ArangoDB client.
#client = ArangoClient(protocol='https', 
#                      host='internal-a43b7a17d62bd11e9aff20eb64c4be3d-1199645507.us-east-1.elb.amazonaws.com', 
#                      port=8529,
#                      http_client=CustomHTTPClient()
#                     )

#db = client.db('rsa_chartreview_ibc', username='tianxia_zhao', password='A4?HLjOd32PsZkD')


#### Use the chart_id to get the instances for all charts from link_json and txt folders

In [3]:
chart_id_list=['ibxrma2019_mult_1212888940010001_hmk_5d75rx3ya75_patricia_colajezzi_002806405_2018',
              'ibxrma2019_mult_1196453270010001_hmk_5a88un2pw44_david_reese_002821057_2018']

In [5]:
chart_id_list[0]

'ibxrma2019_mult_1212888940010001_hmk_5d75rx3ya75_patricia_colajezzi_002806405_2018'

In [6]:
import json
import re
from nltk.corpus import stopwords
import nltk
import string
import csv
import pandas as pd
import numpy as np

In [7]:
# function to extract the text 75 char before and after
charsize=75
def create_feature_clean(sub,sub2):
    print ("inside create feature clean")
    col=['chart_id', 'code', 'text', 'start_offset','end_offset',
        'text75']
    testDF=pd.DataFrame(columns=col)
    #text=sub['chart_text']
    text=sub2
    for e in sub['encounters']:
        for d in e['diagnosisCode']:
            testDF.loc[len(testDF)]=[sub['chartId'], d['code'], d['text'],d['start'], d['end'],text[d['start']-charsize:d['start']+len(d['text'])+charsize]]
    
    # apply some basic processing or noise cleaning
    
    nltk.download('stopwords')
    stop = stopwords.words('english')
    
    testDF['text75'] = testDF['text75'].apply(lambda x: x.lower())
    testDF['text75'] = testDF['text75'].apply(lambda x: x.translate(str.maketrans('','',string.punctuation)))
    # remove string punctuations from text, and remove stop word
    testDF['text75']= testDF['text75'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop)]))
    testDF['text75']=testDF['text75'].str.findall('\w{2,}').str.join(' ')
    return testDF

In [26]:
 # use link_json  and txt files to get data
testDF=pd.DataFrame()
for i in range(len(chart_id_list)):
    chartid=chart_id_list[i]
    # get link json
    content_object= s3resource.Object('chartai-demo-qa','familyhistory_text/'+chartid+'_link.json')  
    file_content = content_object.get()['Body'].read().decode('utf-8')
    datajson = json.loads(file_content)
    # download txt file   Maybe there's a simpler way, not need to 
    content_object= s3resource.Object('chartai-demo-qa','familyhistory_text/'+chartid+'.txt')  #,'txt/'+chartid+'_link.json' )
    datatext = content_object.get()['Body'].read().decode('utf-8')
    #data= s3client.download_file('chartai-demo-qa','familyhistory_text/'+chartid+'.txt','txt/'+chartid+'.txt'  )
    #datatext=open('txt/'+chartid+'.txt','r').read().replace('\n',' ')  # now save to datatext
   # os.remove('txt/'+chartid+'.txt')        # remove the temp folder and file which saved the text file
    # call the function to extract the 75 characters
    testDF1=create_feature_clean(datajson,datatext)
    testDF=pd.concat([testDF,testDF1])    # save all into one


inside create feature clean


[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


inside create feature clean


[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [27]:
len(testDF)   # now it's more than 188444 instances from coder feedback only

531

In [28]:
testDF['chart_id'].nunique()   # coder feedback has 7654 chart_id, somewhere the offset mismatch or something got us only 7502 charts left

2

In [29]:
testDF.columns

Index(['chart_id', 'code', 'text', 'start_offset', 'end_offset', 'text75'], dtype='object')

In [31]:
testDF.to_csv('bfAI_allinst_2charts_062019.csv')

#### load the model and run on these 531 instances

In [7]:
!pip install boto3

    100% |████████████████████████████████| 133kB 18.0MB/s ta 0:00:01
    100% |████████████████████████████████| 71kB 23.9MB/s ta 0:00:01
    100% |████████████████████████████████| 5.6MB 10.9MB/s ta 0:00:01
    100% |████████████████████████████████| 552kB 37.2MB/s ta 0:00:01


In [8]:
import csv
import pandas as pd
import numpy as np
import seaborn as sns
import re
from nltk.corpus import stopwords
import nltk
import string
import matplotlib.pyplot as plt
import random
%matplotlib inline
import os

from sklearn import model_selection, preprocessing, linear_model, naive_bayes, metrics, svm
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer, TfidfTransformer
from sklearn import decomposition, ensemble
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import cross_val_score


from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
from sklearn.multiclass import OneVsRestClassifier

from sklearn.metrics import classification_report

from sklearn.metrics import confusion_matrix

import joblib
import os,boto3,sys,glob,json

In [9]:
testDF=pd.read_csv('bfAI_allinst_2charts_062019.csv')

In [10]:
testDF = testDF[pd.notnull(testDF['text'])] 
testDF = testDF[pd.notnull(testDF['text75'])] 
len(testDF)  # remove all the null text instances

531

In [11]:
testDF=testDF.reset_index(drop=True)
testDF.tail()


Unnamed: 0                                           chart_id   code  \
526         146  ibxrma2019_mult_1196453270010001_hmk_5a88un2pw...  E1165   
527         147  ibxrma2019_mult_1196453270010001_hmk_5a88un2pw...   E119   
528         148  ibxrma2019_mult_1196453270010001_hmk_5a88un2pw...   Z794   
529         149  ibxrma2019_mult_1196453270010001_hmk_5a88un2pw...   E119   
530         150  ibxrma2019_mult_1196453270010001_hmk_5a88un2pw...    I10   

                         text  start_offset  end_offset  \
526  type 2 diabetes mellitus         61395       61419   
527           type 2 diabetes         61966       61981   
528                   levemir         62235       62242   
529  diabetes mellitus type 2         62601       62625   
530    essential hypertension         62691       62713   

                                                text75  
526  management assessment diagnosis descriptioncod...  
527  dateprovider namenotes followup4 monthsstreisf...  
528  ter based blood sugar pattern going increase m...  
529  list descriptiondiagnosi onsetresolutio provid...  
530  us type disorder 4405400 1218201streisfeld nei...

In [12]:
testDF['chart_id'].nunique()

2

In [13]:
# load the model
#files=glob.glob("../picklefolder_ngrams/refreshed_models/*.pickle")
files=glob.glob("../picklefolder_ngrams/pipeline_pickles/*.pickle")

In [14]:
files

['../picklefolder_ngrams/pipeline_pickles/Code_Cholesterol_5Mar.pickle',
 '../picklefolder_ngrams/pipeline_pickles/DeleteReason_FamilyHistory_Mar5.pickle',
 '../picklefolder_ngrams/pipeline_pickles/DeletionReason_Negation_Feb26.pickle',
 '../picklefolder_ngrams/pipeline_pickles/DeletionReason_DoctorsNote_Feb27.pickle',
 '../picklefolder_ngrams/pipeline_pickles/Code_Model2_Mar5.pickle',
 '../picklefolder_ngrams/pipeline_pickles/DeletionReason_NotRelevant_Feb27.pickle',
 '../picklefolder_ngrams/pipeline_pickles/Phrase_Stroke_Feb1819.pickle',
 '../picklefolder_ngrams/pipeline_pickles/Phrase_Hypertension_Feb1819.pickle',
 '../picklefolder_ngrams/pipeline_pickles/Phrase_Emphysema_Mar5.pickle',
 '../picklefolder_ngrams/pipeline_pickles/Phrase_Diabetes_Feb1819.pickle',
 '../picklefolder_ngrams/pipeline_pickles/Phrase_Depression_Mar5.pickle',
 '../picklefolder_ngrams/pipeline_pickles/Phrase_Dementia_Mar5.pickle',
 '../picklefolder_ngrams/pipeline_pickles/Phrase_Ckd_Mar5.pickle',
 '../picklefol

In [15]:
files[1]

'../picklefolder_ngrams/pipeline_pickles/DeleteReason_FamilyHistory_Mar5.pickle'

In [16]:
allpredict=pd.DataFrame()   # create an empty dataframe, to save the prediction result for each model
allprob=pd.DataFrame()
print(allprob)

Empty DataFrame
Columns: []
Index: []


In [17]:
# for one model evaluation

indexdf=pd.DataFrame() 
model_pipeline = joblib.load(files[22])
  #  f = open("allmodeltestscore_0506.txt", "a+")
    # deletion model

testDF1=testDF
indexdf[0]= [1 for i in range(len(testDF))]
        
        
        #print(files[i])
    # how to put the prediction on the right index?
test=testDF1['text75']
print("model loaded:", files )
print("length of data for the model:", len(test))
    
   
proba = model_pipeline.predict_proba(test)[:,1]
prediction = np.where(proba > 0.5, 1, 0) # use a lower threshold
testDF1['key']=testDF1.index
testDF1['pred']=prediction
testDF1['prob']= proba

indexdf['key']=indexdf.index
merged=pd.merge(indexdf,testDF1,on='key',how='outer')  # maintain the position of the perdiction, match to the right index
allpredict[0]=merged['pred']    # save the prediction to corresponding dataframe
allprob[0]=merged['prob']


model loaded: ['../picklefolder_ngrams/pipeline_pickles/Code_Cholesterol_5Mar.pickle', '../picklefolder_ngrams/pipeline_pickles/DeleteReason_FamilyHistory_Mar5.pickle', '../picklefolder_ngrams/pipeline_pickles/DeletionReason_Negation_Feb26.pickle', '../picklefolder_ngrams/pipeline_pickles/DeletionReason_DoctorsNote_Feb27.pickle', '../picklefolder_ngrams/pipeline_pickles/Code_Model2_Mar5.pickle', '../picklefolder_ngrams/pipeline_pickles/DeletionReason_NotRelevant_Feb27.pickle', '../picklefolder_ngrams/pipeline_pickles/Phrase_Stroke_Feb1819.pickle', '../picklefolder_ngrams/pipeline_pickles/Phrase_Hypertension_Feb1819.pickle', '../picklefolder_ngrams/pipeline_pickles/Phrase_Emphysema_Mar5.pickle', '../picklefolder_ngrams/pipeline_pickles/Phrase_Diabetes_Feb1819.pickle', '../picklefolder_ngrams/pipeline_pickles/Phrase_Depression_Mar5.pickle', '../picklefolder_ngrams/pipeline_pickles/Phrase_Dementia_Mar5.pickle', '../picklefolder_ngrams/pipeline_pickles/Phrase_Ckd_Mar5.pickle', '../picklefo

/opt/conda/lib/python3.6/site-packages/sklearn/base.py:251: UserWarning: Trying to unpickle estimator CountVectorizer from version 0.18.1 when using version 0.20.2. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/base.py:251: UserWarning: Trying to unpickle estimator LogisticRegression from version 0.18.1 when using version 0.20.2. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/base.py:251: UserWarning: Trying to unpickle estimator LabelBinarizer from version 0.18.1 when using version 0.20.2. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/base.py:251: UserWarning: Trying to unpickle estimator LogisticRegression from version 0.18.1 when using version 0.20.2. This might lead to breaking code or invalid results. Use at your 

In [18]:
len(allpredict)

531

In [19]:
allpredict.isnull().sum()   # all instances are run

0    0
dtype: int64

In [20]:
# for one model, this step is not necessary, because "pred" column already has the info already
testDF['index']=testDF.index
allpredict['index']=allpredict.index
df_pred = pd.merge(testDF,allpredict, how='left', on = 'index')
len(df_pred)

531

In [21]:
df_pred.tail()

Unnamed: 0                                           chart_id   code  \
526         146  ibxrma2019_mult_1196453270010001_hmk_5a88un2pw...  E1165   
527         147  ibxrma2019_mult_1196453270010001_hmk_5a88un2pw...   E119   
528         148  ibxrma2019_mult_1196453270010001_hmk_5a88un2pw...   Z794   
529         149  ibxrma2019_mult_1196453270010001_hmk_5a88un2pw...   E119   
530         150  ibxrma2019_mult_1196453270010001_hmk_5a88un2pw...    I10   

                         text  start_offset  end_offset  \
526  type 2 diabetes mellitus         61395       61419   
527           type 2 diabetes         61966       61981   
528                   levemir         62235       62242   
529  diabetes mellitus type 2         62601       62625   
530    essential hypertension         62691       62713   

                                                text75  key  pred      prob  \
526  management assessment diagnosis descriptioncod...  526     0  0.194300   
527  dateprovider namenotes followup4 monthsstreisf...  527     0  0.423412   
528  ter based blood sugar pattern going increase m...  528     1  0.570587   
529  list descriptiondiagnosi onsetresolutio provid...  529     0  0.200042   
530  us type disorder 4405400 1218201streisfeld nei...  530     1  0.692155   

     index  0  
526    526  0  
527    527  0  
528    528  1  
529    529  0  
530    530  1

In [22]:
df_pred['pred'].value_counts()  # 37 predicted as delete

1    311
0    220
Name: pred, dtype: int64

In [23]:
len(df_pred[df_pred['text75'].str.contains('family history',flags=re.IGNORECASE)])  # check all 

36

In [24]:
del_sub=df_pred[df_pred['pred']==1]

In [25]:
len(del_sub[del_sub['text75'].str.contains('family history',flags=re.IGNORECASE)]) # check deleted instances if text75 contains "family history"

35

In [ ]:
pd.options.display.max_colwidth = 250 # make it show the complete string

In [58]:
print(del_sub[['text75']])

                                                                                                                                                    text75
51               history family history cardiac disorder z8249 family history diabetes mellitus z833 family history hypertension z8249 family history mali
52                      z8249 family history diabetes mellitus z833 family history hypertension z8249 family history malignant neoplasm breast z803 social
53         itus z833 family history hypertension z8249 family history malignant neoplasm breast z803 social history never smoked substance z789 never smok
58   yroid surgery wedge resection lung never smoker family history malignant neoplasm breast family history hypertension family history diabetes mellitus
59               moker family history malignant neoplasm breast family history hypertension family history diabetes mellitus alcohol use lung wedge resect
60         nt neoplasm breast family history hypertension family histo

In [57]:
# this is the one deleted but without "family history" in text75

print(del_sub['text75'][del_sub['text75'].str.contains('family history',flags=re.IGNORECASE)==False])

512    cva cerebral vascular accident diabetes mellitus type controlled colon cancer screening prostate cancer screening enlarged prostate without lower ur
Name: text75, dtype: object
